src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+    >>>     for index in asignment.lhs.children:
+    >>>         try:
+    >>>             trans.apply(index)
+    >>>         except TransformationError:
+    >>>             pass
+    >>> schedule.view()
+
+    '''
+    def apply(self, node, options=None):
+        '''Apply the NemoArrayAccess2Loop transformation if the supplied node
+        is an access to an array index within an Array Reference that
+        is on the left-hand-side of an Assignment node. The access
+        must be a scalar (i.e. not a range) and must not include a
+        loop variable (as we are transforming a single access to a
+        loop).
+
+        These constraints are required for correctness and an
+        exception will be raised if they are not satisfied. If the
+        constraints are satisfied then the array access is replaced
+        with a loop iterator and a single trip loop.
+
+        If there are existing loops then the new loop will be placed
+        at the expected nesting, complying to the ordering expected by
+        the NEMO API (specified in the config file). If the existing
+        loops do not conform to this expected ordering then an
+        exception will be raised.
+
+        The name of the loop index is taken from the PSyclone
+        configuration file if a name exists for the particular array
+        index, otherwise a new name is generated.
+
+        :param node: an array index.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        '''
+        self.validate(node)
+
+        array_index = node.position
+        array_reference = node.parent
+        assignment = array_reference.parent
+        symbol_table = node.scope.symbol_table
+
+        node_copy = node.copy()
+
+        # See if there is any configuration information for this array index
+        loop_type_order = Config.get().api_conf("nemo").get_index_order()
+        # TODO: Add tests in get_loop_type_data() to make sure values
+        # are strings that represent an integer or a valid variable
+        # name, e.g. 1a should not be allowed. See issue #1035
+        loop_type_data = Config.get().api_conf("nemo").get_loop_type_data()
+        try:
+            loop_type = loop_type_order[array_index]
+            loop_type_info = loop_type_data[loop_type]
+            loop_variable_name = loop_type_info['var']
+        except IndexError:
+            loop_variable_name = symbol_table.next_available_name("idx")
+
+        # Work out where to add the new loop (at 'location'), as there
+        # may be existing loops that should be inside this one.
+        symbols = [ref.symbol for ref in array_reference.children[:array_index]
+                   if isinstance(ref, Reference)]
```

Comment:
This doesn't look very robust to me, e.g. what about:

    jtop = 10
    do ji = 1, jpi
      array(ji+1, jtop, jk) = xxx`
    end do

If possible, I think I would find all loops that enclose the current assignment (maybe search up to an instance of `Routine`?), find the loop variable for each of them and then search the array index expressions to see if they are present - probably using @hiker's dependence analysis tools. Alternatively, since we're introducing a 1-trip loop, we could just introduce it at the innermost level (i.e. immediately around the assignment)? Moving it can be done subsequently by e.g. @hiker's LoopSwap transformation or in fact, we leave it to the the SIR backend.

Reply:
This has been removed in this PR as it no longer attempt to place loops in the ordering expected by NEMO. As suggested by the reviewer, this is being deferred to a separate issue - see #1328.